### PR TITLE
Add C implementation for `@stdlib/math/base/special/log1p`

### LINKED ISSUE #641: [RFC]: Add C implementation for `@stdlib/math/base/special/log1p`
### Description

This RFC proposes adding a C implementation for [`@stdlib/math/base/special/log1p`](https://github.com/stdlib-js/stdlib/tree/b43a75923e47775446d3b6d7219ee42e07f7ce63/lib/node_modules/%40stdlib/math/base/special/log1p).

```c
double stdlib_base_log1p( const double x );
```

### Related Issues

None.

### Questions

No.

### Other

No.

### Checklist

- [X] I have read and understood the [Code of Conduct](https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md).
- [X] Searched for existing issues and pull requests.
- [X] The issue name begins with `RFC:`.

### PR DESCRIPTION
**Pull request by Pranavchiku**
Resolves #641.

## Checklist

- [x] update readme.md
- [x] include.gypi
- [x] binding.gyp
- [x] include/stdlib/math/base/special/
- [x] evalpoly.js
- [x] src
- [x] manifest.json
- [x] lib
- [x] examples
- [x] benchmark
- [x] test
* * *

@kgryte

### CODE REVIEW

**Comment by Pranavchiku** [comment]
All tests and benchmarks cleared, this PR can have a final review.

File: lib/node_modules/@stdlib/math/base/special/log1p/README.md [hunk 1]
```diff
@@ -89,6 +89,96 @@ for ( i = 0; i < 100; i++ ) {
...
+#include <stdlib.h>
+#include <stdio.h>
+
+int main() {
+    double x;
+    double v;
+    int i;
+    
+    for ( i = 0; i < 100; i++ ) {
+        x = (double)rand() / (double)RAND_MAX * 100.0;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/log1p/README.md]:
```suggestion
        x = ( (double)rand() / (double)RAND_MAX ) * 100.0;
```

File: lib/node_modules/@stdlib/math/base/special/log1p/benchmark/c/native/benchmark.c [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+int main( void ) {
+	double elapsed;
+	int i;
+
+	// Use the current time to seed the random number generator:
+	srand( time( NULL ) );
+
+	print_version();
+	for ( i = 0; i < REPEATS; i++ ) {
+		printf( "# c::%s\n", NAME );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/log1p/benchmark/c/native/benchmark.c]:
```suggestion
		printf( "# c::native::%s\n", NAME );
```

File: lib/node_modules/@stdlib/math/base/special/log1p/examples/c/example.c [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+#include <stdlib.h>
+#include <stdio.h>
+
+int main() {
+    double x;
+    double v;
+    int i;
+
+    for ( i = 0; i < 100; i++ ) {
+        x = (double)rand() / (double)RAND_MAX * 100.0;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/log1p/examples/c/example.c]:
```suggestion
        x = ( (double)rand() / (double)RAND_MAX ) * 100.0;
```

File: lib/node_modules/@stdlib/math/base/special/log1p/include/stdlib/math/base/special/log1p.h [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+/*
+* If C++, prevent name mangling so that the compiler emits a binary file having undecorated names, thus mirroring the behavior of a C compiler.
+*/
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+* Evaluates the natural logarithm of 1+x.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/log1p/include/stdlib/math/base/special/log1p.h]:
```suggestion
* Evaluates the natural logarithm of `1+x`.
```

File: lib/node_modules/@stdlib/math/base/special/log1p/lib/native.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+var addon = require( './../src/addon.node' );
+
+
+// MAIN //
+
+/**
+* Evaluates the natural logarithm of \\(1+x\\).
+*
+* @param {number} x - input value
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/log1p/lib/native.js:32]:
```suggestion
* @private
* @param {number} x - input value
```

File: lib/node_modules/@stdlib/math/base/special/log1p/src/main.c [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+* [horners-method]: https://en.wikipedia.org/wiki/Horner%27s_method
+*
+* @param x    value at which to evaluate the polynomial
+* @return     evaluated polynomial
+*/
+static double polyval_lp( const double x ) {
+	return 0.6666666666666735 + (x * (0.3999999999940942 + (x * (0.2857142874366239 + (x * (0.22222198432149784 + (x * (0.1818357216161805 + (x * (0.15313837699209373 + (x * 0.14798198605116586)))))))))));
+}
+
+// END: polyval_lp
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/log1p/src/main.c:83]:
```suggestion
// END: polyval_lp

/* End auto-generated functions. */
```

File: lib/node_modules/@stdlib/math/base/special/log1p/src/main.c [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+
+/**
+* Evaluates the natural logarithm of 1+x.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/log1p/src/main.c]:
```suggestion
* Evaluates the natural logarithm of `1+x`.
```

File: lib/node_modules/@stdlib/math/base/special/log1p/src/main.c [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+* @return	  output value
+*
+* @example
+* double out = stdlib_base_log1p( 4.0 );
+* // returns ~1.609
+*/
+double stdlib_base_log1p( const double x ) {
+	double hfsq;
+	double xc = x;
+	uint32_t hu;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/log1p/src/main.c:84]:
```suggestion
	uint32_t hu;
	double xc;
```

File: lib/node_modules/@stdlib/math/base/special/log1p/src/main.c [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+	double y;
+	double f;
+	double c;
+	double s;
+	double z;
+	double R;
+	double u;
+	int k;
+
+	if ( xc < -1.0 || stdlib_base_is_nan( xc ) ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/log1p/src/main.c:94]:
```suggestion
	if ( x < -1.0 || stdlib_base_is_nan( x ) ) {
```

File: lib/node_modules/@stdlib/math/base/special/log1p/src/main.c [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+		return 0.0 / 0.0; // NaN
+	}
+	if ( xc == -1.0 ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/log1p/src/main.c:97]:
```suggestion
	if ( x == -1.0 ) {
```

File: lib/node_modules/@stdlib/math/base/special/log1p/src/main.c [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+		return STDLIB_CONSTANT_FLOAT64_NINF;
+	}
+	if ( xc == STDLIB_CONSTANT_FLOAT64_PINF ) {
+		return xc;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/log1p/src/main.c:101]:
```suggestion
	if ( x == STDLIB_CONSTANT_FLOAT64_PINF ) {
		return x;
```

**Review by kgryte — APPROVED** [review]

**Review by kgryte — APPROVED** [review]

**Comment by kgryte** [comment]
Thanks, @Pranavchiku!

**Pull request merged**